File: packages/protocol/test/vectors.test.ts
// Cross-language conformance: reproduces every value in
// fixtures/commitments/vectors.json, the same fixture the Rust suite in
// crates/commitments checks against. A mismatch here means the two
// languages would produce different commitments for identical inputs.

import { describe, expect, it } from "vitest";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import {
  canonicalizeValue,
  frameHash,
  dataCommitment,
  leafHash,
  MerkleTree,
  verifyInclusion,
} from "../src/index.js";

const fixturePath = fileURLToPath(
  new URL("../../../fixtures/commitments/vectors.json", import.meta.url),
);
const vectors = JSON.parse(readFileSync(fixturePath, "utf8"));

function hex32(s: string): Uint8Array {
  const bytes = new Uint8Array(s.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(s.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

describe("canonical encoding", () => {
  it("matches the fixture", () => {
    const actual = canonicalizeValue(vectors.canonical.input);
    expect(new TextDecoder().decode(actual)).toBe(vectors.canonical.expectedBytesUtf8);
  });
});

describe("frame hash", () => {
  it("matches the fixture", () => {
    const fields: Uint8Array[] = vectors.frameHash.fieldsUtf8.map(
      (f: string) => new TextEncoder().encode(f),
    );
    const actual = frameHash(vectors.frameHash.tag, fields);
    expect(toHex(actual)).toBe(vectors.frameHash.expectedHex);
  });
});

describe("data commitment", () => {
  it("matches the fixture and reacts to salt/field changes", () => {
    const data = vectors.dataCommitment.data;
    const commitments: Uint8Array[] = [];
    for (const c of vectors.dataCommitment.cases) {
      const salt = hex32(c.saltHex);
      const actual = dataCommitment(salt, data);
      expect(toHex(actual), `case ${c.name}`).toBe(c.expectedHex);
      commitments.push(actual);
    }
    expect(toHex(commitments[0]!)).not.toBe(toHex(commitments[1]!));

    const fieldChanged = vectors.dataCommitment.fieldChangedCase;
    const actual = dataCommitment(hex32(fieldChanged.saltHex), fieldChanged.data);
    expect(toHex(actual)).toBe(fieldChanged.expectedHex);
    expect(toHex(actual)).not.toBe(toHex(commitments[0]!));
  });
});

describe("merkle tree", () => {
  it("matches the fixture end to end", () => {
    const m = vectors.merkle;
    const leaves: Uint8Array[] = m.leafRecords.map((record: unknown, i: number) =>
      leafHash(hex32(m.leafSaltsHex[i]), record),
    );
    const expectedLeafHashes: string[] = m.expectedLeafHashesHex;
    expect(leaves.map(toHex)).toEqual(expectedLeafHashes);

    const tree = MerkleTree.build(leaves);
    expect(toHex(tree.root)).toBe(m.expectedRootHex);
    expect(tree.height()).toBe(m.expectedHeight);

    const proofFixture = m.inclusionProof;
    const proof = tree.prove(proofFixture.index);
    expect(proof.siblings.map(toHex)).toEqual(proofFixture.expectedSiblingsHex);
    expect(proof.count).toBe(m.leafRecords.length);

    expect(() => verifyInclusion(leaves[proofFixture.index]!, proof, tree.root)).not.toThrow();

    const emptyTree = MerkleTree.build([]);
    expect(toHex(emptyTree.root)).toBe(m.emptyTree.expectedRootHex);

    const singleTree = MerkleTree.build([leaves[0]!]);
    expect(toHex(singleTree.root)).toBe(m.singleLeafTree.expectedRootHex);
    expect(singleTree.height()).toBe(m.singleLeafTree.expectedHeight);
  });
});

File: crates/commitments/tests/vectors_test.rs
//! Cross-language conformance: reproduces every value in
//! `fixtures/commitments/vectors.json`, the same fixture the TypeScript
//! suite in `packages/protocol` checks against. A mismatch here means the
//! two languages would produce different commitments for identical inputs.

use commitments::*;
use serde_json::Value;
use std::fs;

fn vectors() -> Value {
    let path = concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../fixtures/commitments/vectors.json"
    );
    let text = fs::read_to_string(path).expect("fixture file must be readable");
    serde_json::from_str(&text).expect("fixture file must be valid JSON")
}

fn hex32(s: &str) -> [u8; 32] {
    let bytes = hex::decode(s).expect("valid hex");
    bytes.try_into().expect("32 bytes")
}

#[test]
fn canonical_encoding_matches_fixture() {
    let v = vectors();
    let input = &v["canonical"]["input"];
    let expected = v["canonical"]["expectedBytesUtf8"].as_str().unwrap();
    let actual = canonicalize_value(input).unwrap();
    assert_eq!(String::from_utf8(actual).unwrap(), expected);
}

#[test]
fn frame_hash_matches_fixture() {
    let v = vectors();
    let tag = v["frameHash"]["tag"].as_str().unwrap();
    let fields: Vec<Vec<u8>> = v["frameHash"]["fieldsUtf8"]
        .as_array()
        .unwrap()
        .iter()
        .map(|f| f.as_str().unwrap().as_bytes().to_vec())
        .collect();
    let field_refs: Vec<&[u8]> = fields.iter().map(|f| f.as_slice()).collect();
    let expected = hex32(v["frameHash"]["expectedHex"].as_str().unwrap());
    assert_eq!(frame_hash(tag, &field_refs).unwrap(), expected);
}

#[test]
fn data_commitment_matches_fixture_and_reacts_to_salt_and_field_changes() {
    let v = vectors();
    let data = &v["dataCommitment"]["data"];
    let cases = v["dataCommitment"]["cases"].as_array().unwrap();

    let mut commitments = Vec::new();
    for case in cases {
        let salt = hex32(case["saltHex"].as_str().unwrap());
        let expected = hex32(case["expectedHex"].as_str().unwrap());
        let actual = data_commitment(&salt, data).unwrap();
        assert_eq!(actual, expected, "case {}", case["name"]);
        commitments.push(actual);
    }
    // Changing only the salt must change the commitment.
    assert_ne!(commitments[0], commitments[1]);

    let field_changed = &v["dataCommitment"]["fieldChangedCase"];
    let salt = hex32(field_changed["saltHex"].as_str().unwrap());
    let changed_data = &field_changed["data"];
    let expected = hex32(field_changed["expectedHex"].as_str().unwrap());
    let actual = data_commitment(&salt, changed_data).unwrap();
    assert_eq!(actual, expected);
    // Changing only a field (same salt as the baseline case) must change
    // the commitment too.
    assert_ne!(actual, commitments[0]);
}

#[test]
fn merkle_tree_matches_fixture() {
    let v = vectors();
    let m = &v["merkle"];
    let records = m["leafRecords"].as_array().unwrap();
    let salts_hex = m["leafSaltsHex"].as_array().unwrap();

    let leaves: Vec<[u8; 32]> = records
        .iter()
        .zip(salts_hex)
        .map(|(record, salt_hex)| {
            let salt = hex32(salt_hex.as_str().unwrap());
            leaf_hash(&salt, record).unwrap()
        })
        .collect();

    let expected_leaf_hashes: Vec<[u8; 32]> = m["expectedLeafHashesHex"]
        .as_array()
        .unwrap()
        .iter()
        .map(|h| hex32(h.as_str().unwrap()))
        .collect();
    assert_eq!(leaves, expected_leaf_hashes);

    let tree = MerkleTree::build(leaves.clone()).unwrap();
    assert_eq!(tree.root, hex32(m["expectedRootHex"].as_str().unwrap()));
    assert_eq!(tree.height() as u64, m["expectedHeight"].as_u64().unwrap());

    let proof_fixture = &m["inclusionProof"];
    let index = proof_fixture["index"].as_u64().unwrap();
    let proof = tree.prove(index).unwrap();
    let expected_siblings: Vec<[u8; 32]> = proof_fixture["expectedSiblingsHex"]
        .as_array()
        .unwrap()
        .iter()
        .map(|h| hex32(h.as_str().unwrap()))
        .collect();
    assert_eq!(proof.siblings, expected_siblings);
    assert_eq!(
        proof.count,
        m["leafRecords"].as_array().unwrap().len() as u64
    );

    verify_inclusion(&leaves[index as usize], &proof, &tree.root).expect("valid proof must verify");

    let empty_tree = MerkleTree::build(vec![]).unwrap();
    assert_eq!(
        empty_tree.root,
        hex32(m["emptyTree"]["expectedRootHex"].as_str().unwrap())
    );

    let single_tree = MerkleTree::build(vec![leaves[0]]).unwrap();
    assert_eq!(
        single_tree.root,
        hex32(m["singleLeafTree"]["expectedRootHex"].as_str().unwrap())
    );
    assert_eq!(
        single_tree.height() as u64,
        m["singleLeafTree"]["expectedHeight"].as_u64().unwrap()
    );
}
